### PR TITLE
Parallel execution of containerized evaluations

### PR DESCRIPTION
**Pull request by Munsio**

### CODE REVIEW

**Comment by Munsio** [comment]
@bauersimon I would like to have your opinion on this one.

**Comment by bauersimon** [comment]
plz rebase and then tag me again

**Comment by Munsio** [comment]
Alright I got it to work. I also updated the follow up issue already to also add tests for `--parallel` when we run docker tests on the CI.

Please check.

Also if you have a idea for testing the `Parallel` stuff please tell me.

Also here is the command I used to run:
```

File: util/exec.go [hunk 1]
```diff
@@ -90,3 +91,43 @@ func FilterArgs(args []string, ignored []string) (filtered []string) {
...
 }
+
+// Parallel holds a buffered channel for limiting parallel executions.
+type Parallel struct {
+	ch chan struct{}
+	wg sync.WaitGroup
+}
+
+// NewParallel returns a Parallel execution limiting helper.
+func NewParallel(limit uint, timeout time.Duration) *Parallel {
```
Comment (bauersimon) [inline util/exec.go:102]: Would maybe call that `maxWorkers`.

File: util/exec_test.go [hunk 2]
```diff
@@ -91,3 +91,39 @@ func TestFilterArgs(t *testing.T) {
...
+		Name: "Run 10 times with limit of 2",
+
+		Timeout: 0,
+		Limit:   2,
+		Times:   10,
+
+		Function: func() {
+			// do something?
+		},
+	})
```
Comment (bauersimon) [inline util/exec_test.go:128]: Would have one test case where every function adds `+1` to some mutex and then you check if really all 10x of them ran.
And a second case maybe where you have `time.Sleep(1)` in every function with `Limit:1` and `Times:3` and make sure the whole thing runs for at least like `2.5` seconds.

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 3]
```diff
@@ -74,6 +74,8 @@ type Evaluate struct {
 	Runtime string `long:"runtime" description:"The runtime which will be used for the evaluation." default:"local" choice:"local" choice:"docker"`
 	// RuntimeImage determines the container image used for any container runtime.
 	RuntimeImage string `long:"runtime-image" description:"The container image to use for the evaluation." default:""`
+	// Parallel holds the number of parallel executed runs.
+	Parallel uint `long:"parallel" description:"Amount of parallel executed runs." default:"1"`
```
Comment (bauersimon) [inline cmd/eval-dev-quality/cmd/evaluate.go:78]: State that this only affects docker.

File: util/exec.go (hunk 1 above)
Comment (bauersimon) [inline util/exec.go:102]: Timeout parameter unused.

File: cmd/eval-dev-quality/cmd/evaluate.go (hunk 3 above)
Comment (bauersimon) [reply cmd/eval-dev-quality/cmd/evaluate.go:78]: Though I would maybe make the logic different (maybe in a follow up). Cause now we have runs and parallel executions... Would be cool if `parallel` would be a boolean and it automatically distributes multiple runs across parallel executions?

**Review by bauersimon — CHANGES_REQUESTED** [review]

File: cmd/eval-dev-quality/cmd/evaluate.go (hunk 3 above)
Comment (Munsio) [reply cmd/eval-dev-quality/cmd/evaluate.go:78]: Yeah but runs and parallel are different.
For better result spread I want to have 5 runs of each model, while my server is able to run 15-20 parallel evaluations.

Making it exclusive for containerized runs and clarify it in the comment is the way to go I guess.

File: util/exec_test.go (hunk 2 above)
Comment (Munsio) [reply util/exec_test.go:128]: The first one sounds right and enough, I do not want to use timed counts with `sleep` as they artificially increase execution time where it is not needed, as in fact both do the same thing only the measurement is different.

File: util/exec.go (hunk 1 above)
Comment (Munsio) [reply util/exec.go:102]: Ah yeah forgot to remove, thanks.

**Comment by Munsio** [comment]
@bauersimon please review again.

File: util/exec_test.go (hunk 2 above)
Comment (bauersimon) [reply util/exec_test.go]: mhmmm... if you don't want sleeps than I am not sure how to easily measure that only x run at a time.

File: util/exec_test.go [hunk 4]
```diff
@@ -91,3 +92,35 @@ func TestFilterArgs(t *testing.T) {
...
+
+		p := NewParallel(tc.Limit)
+		for i := uint(0); i < tc.Times; i++ {
+			p.Execute(tc.Function)
+		}
+		p.Wait()
+
+		assert.Equal(t, count.Load(), uint32(expectedCount))
+	})
+
```
Comment (bauersimon) [inline util/exec_test.go:125]: Unused newline.

File: util/exec_test.go [hunk 5]
```diff
@@ -91,3 +92,35 @@ func TestFilterArgs(t *testing.T) {
...
 	})
 }
+
+func TestParallelExecute(t *testing.T) {
+	type testCase struct {
+		Limit uint
+		Times uint
+
+		Function func()
+	}
```
Comment (bauersimon) [inline util/exec_test.go:102]: Kinda weird having table tests that are not used in any table fashion. Would drop them.

File: util/exec_test.go [hunk 6]
```diff
@@ -91,3 +92,35 @@ func TestFilterArgs(t *testing.T) {
...
+
+	t.Run("Atomic count", func(t *testing.T) {
```
Comment (bauersimon) [inline util/exec_test.go:104]: So this ensures that the functions are executed correctly. Not sure how to test that the limit works without using `time.Sleep`. I guess you could make something that adds and subtracts from atomic and also checks what the maximum value was. And with a limit of 1 then the maximum value should never exceed 2. But as long as it is functionally correct, we are good I think.

**Review by bauersimon — DISMISSED** [review]

File: util/exec_test.go (hunk 2 above)
Comment (Munsio) [reply util/exec_test.go]: Hmm yeah you are right, testing the limit is only possible with time. Fixed that up. Please review again @bauersimon

File: util/exec_test.go (hunk 5 above)
Comment (bauersimon) [reply util/exec_test.go:102]: Still weird cause we construct them within the tests and then just use their properties. So would either introduce a more elaborate `validate` or really just remove the `testCase`. To me it makes everything harder to read.

File: cmd/eval-dev-quality/cmd/evaluate.go [hunk 7]
```diff
@@ -144,6 +146,10 @@ func (command *Evaluate) Initialize(args []string) (evaluationContext *evaluate.
 			}
 		}
 
+		if command.Parallel != 1 && command.Runtime == "local" {
+			command.logger.Panic("the 'parallel' parameter can't be used with local execution")
+		}
+
```
Comment (bauersimon) [inline cmd/eval-dev-quality/cmd/evaluate.go:152]: `command.Parallel` could be `0` ... sry for not noticing earlier

**Comment by Munsio** [comment]
As discussed with @bauersimon everything is ready to merge with the fix of the pipeline.

Merge it right away

**Pull request merged**